\documentclass[11pt,a4paper]{article}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{authblk}
\usepackage{color}
\usepackage{fullpage}

\usepackage{caption}

\usepackage{tikz}
\usetikzlibrary{bayesnet}
\usepackage{graphicx}
\usepackage{amsfonts}
\usepackage{amsmath,bm}
\usepackage{enumitem}
\usepackage[linesnumbered,ruled,vlined]{algorithm2e}

\usepackage{graphicx}
\graphicspath{ {./images/} }

\usepackage{minted}

\usepackage[backend=biber, style=nature, citestyle=nature]{biblatex}

\addbibresource{references.bib} %Imports bibliography file

\renewcommand{\topfraction}{.85}
\renewcommand{\bottomfraction}{.7}
\renewcommand{\textfraction}{.15}
\renewcommand{\floatpagefraction}{.3}
\renewcommand{\dbltopfraction}{.3}
\renewcommand{\dblfloatpagefraction}{.3}
\setcounter{topnumber}{9}
\setcounter{bottomnumber}{9}
\setcounter{totalnumber}{20}
\setcounter{dbltopnumber}{9}

\usepackage{color}
\newcommand{\red}{\textcolor{red}}
\newcommand{\blue}{\textcolor{blue}}

\title{cell2location: high throughput spatial mapping of cell types and their interactions \\
Supplementary Methods
}

\author{Vitalii Kleshchevnikov, Emma Dann, Alexander Aivazidis, Artem Shmatko, Artem Lomakin, Mika Sarkin Jain, Liz Tuck, Anna Arutyunyan, Lauma Ramona, Roser Vento-Tormo, Moritz Gerstung, Oliver Stegle, Omer Bayraktar}
\date{March 2020}

\begin{document}

\maketitle

\tableofcontents

\section{Mapping cell types with cell2location model}

\subsection{Cell2location model} \label{cell2location_model}

Cell2location is a Bayesian model that is aimed at decomposing spot-level expression levels into a set of pre-defined reference signatures of cell types. A representation as graphical model is shown in Fig \ref{fig:graphical_model}. The input data of the method are as follows: \newline

Let $D=\{d_{s,g}\}$ be a $S \times G$ spatial expression count matrix for locations $s=\{1..S\}$ and genes $g=\{1..G\}$. 
For 10X Visium data, this matrix can be directly obtained from the 10X SpaceRanger software (see main text Methods). For other technologies, such as Slide-Seq V2, files with count matrices of the appropriate formats need to be generated by the user. \newline

Let $G=\{g_{f,g}\}$, denote an $F \times G$ matrix of reference signatures of cell types, which consist of $f=\{1..F\}$ reference signatures, each of which is defined by a gene expression profile $G_{f,:}$ for $g=\{1...G\}$ genes. This matrix needs to be provided to cell2location and can be estimated from scRNA-seq profiles. Strategies how to estimate this matrix are presented in Section \ref{c2l_ref_prog}. \newline

Cell2location models the elements of $D$ as Negative Binomial distributed, given an unobserved expression level (rate) $\mu_{s,g}$ and a gene-specific over-dispersion parameter $\alpha_g$ which describes variance in expression of individual genes that is not explained by the reference signatures of cell types: 
\begin{equation} \label{eq:c2l:1}
D_{s,g} \sim \mathtt{NB}(\mu_{s,g}, \alpha_g) \\
\end{equation}

Note that this is equivalent to a Poisson variable (measurement model) with a Gamma-distributed mean (expression model)
\cite{sarkar_separating_2020}.
\begin{equation} \label{eq:c2l:1}
D_{s,g} \sim \mathtt{Poisson}(\mathtt{Gamma}(\alpha_g, \alpha_g / \mu_{s,g})) \\
\end{equation}

The expression of level of genes $\mu_{s,g}$ in the rate space is modelled as a linear additive function of non-negative components:
\begin{equation} \label{eq:c2l:3}
\mu_{s,g} = m_{g} \left (\sum_{f} {w_{s,f} \: g_{f,g}} \right) + l_s + s_{g}\\
\end{equation}

Here, $w_{s,f}$ denotes regression weight of each reference signature $f$ at location $s$; 
$m_{g}$ denotes a gene-specific scaling parameter to account for global differences in expression estimates between technologies;
$l_{s}$ and $s_{g}$ are additive components that capture additive background variation that is not explained by the bi-variate decomposition.

The prior distribution on the unobserved parameters are as follows:

\begin{enumerate}
    \item \textbf{Absolute abundance of cell types across locations.}
    The regression weights $w_{s,f}$ for each reference signature, which represent absolute cell density at a given location, are in turn Gamma distributed (Eq. \ref{eq:c2l:4} - \ref{eq:c2l:11}); and referred to in the cell2location software as \mintinline{python}|'spot_factors'|. \newline
    \begin{equation} \label{eq:c2l:4}
    w_{sf} \sim \mathtt{Gamma}({\mu}^{w}_{sf}, {\mu}^{w}_{sf} / c^{w})
    \end{equation}
    This parameter is modelled using another layer of decomposition to account for linear dependencies between cell type abundance across locations with similar cell composition (note this is not modelling spatial co-variance). \red{this does not account for spatial proximity? It will capture it but the models does not know anything about distance?}
    
    \begin{equation} \label{eq:c2l:5}
    {\mu}^{w}_{sf} = (\sum_{r} {z_{sr} \: x_{rf}})\\
    \end{equation}
    
    The prior mean ${\mu}^{w}_{sf}$ parameter is additively decomposed into the latent groups of cell types $r=\{1...R\}$ to account for linear dependencies in spatial abundance of reference signatures $f$, where $R$ is specified by the user (the default is 50). The scalar constant $c^{w}$ controls the extent to which this co-expression prior controls $\mu^{w}_{s,f}$. The groups of cell types can be thought of as cellular compartments in the tissue. As illustrated by validation on simulated data (Fig S2A), modelling these linear dependencies increases the sensitivity of cell2location for mapping low abundance cell types in particular.
    
    The prior distributions of $z_{sr}$ and $x_{rf}$ are defined to control absolute scale the cell type abundance estimates $w_{sf}$ (the number of cells per location $s$ expressing each cell type reference signature $f$). In addition, these prior distributions control the sparsity of how many cell types $f$ are present in each location $s$, enabling application of cell2location to tissues and technologies with varying numbers of cells and cells types per location.
    
    These priors need to be adjusted by the user to every tissue. We observed 5-6 cell types and 8 cells per location in the mouse brain 10X Visium (Fig S6, Fig S10), but just 1-2 cell types and cells in the human brain and heart (unpublished). In the lymphoid and epithelial tissues, 10X Visium locations contain is 5-10 cell types and 30 cells per location (Fig S10, unpublished). Slide-seq locations have 1-2 cell types and fractions of cells in each location (Fig S10). The model will give more accurate absolute cell abundance estimates when these priors are chosen well.
    
    \red{this is all pretty complicated and I'd appreciate if we can reduce complexity. Can this be simplified / reduced a little?}
    \begin{itemize}
        \item The total spatial abundance $z_{sr}$ of latent cell type groups $r$ across locations $s$ is defined as Gamma distributed with a prior controlling the scale and sparsity of this distribution:
        \begin{equation} \label{eq:c2l:6}
        z_{sr} \sim \mathtt{Gamma}(shape = N_s^{grp} / R, rate = 1 / (N_s^{cells} / N_s^{grp}))
        \end{equation}
    
        where $N_s^{cells}$ is the average number of cells in each location, and $N_s^{grp}$ i the number of latent groups $r$  present in each location $s$ . \newline
        The prior on $z_{sr}$ is specified such that $\sum_{r} z_{sr}$ on average equals to the number of cells per location $\sum_{r} z_{sr} = E(N_s^{cells})$, and that on average each location has high $z_{sr}$ for $N^{grp}$ cell type groups. \newline
        Unobserved $N_s^{grp}$ and $N_s^{cells}$ are modelled as Gamma-distributed variables defined using prior knowledge about the tissue used to produce spatial data $D_{sg}$, in \ref{eq:c2l:7} - \ref{eq:c2l:8}. Note, that the model does not require knowing the exact number of cells obtained by counting nuclei in each spot. \newline
        User needs to provide,   \newline
        \begin{equation} \label{eq:c2l:7}
        N_s^{grp} \sim \mathtt{Gamma}(mean=E(N_s^{grp}), \sigma^2=E(N_s^{grp}) / v^{grp})
        \end{equation}
        \begin{equation} \label{eq:c2l:8}
        N_s^{cells} \sim \mathtt{Gamma}(mean=E(N_s^{cells}), \sigma^2=E(N_s^{cells}) / v^{cells})
        \end{equation}
        where $E(N^{cells-mean})$ is a user-provided rough estimate of the number of cells per location; $N^{grp-mean}$ is a user-provided estimate of the average number of cellular compartments / zones per location; and $v^{grp}$ and $v^{cells}$ are measure of uncertainty in those estimates which are set to $1$.
        When the cell count, obtained by segmenting histology data paired with $D_{s,g}$ for each location, is available it can be provided as a location-specific $E(N_s^{cells})$ prior and $v^{cells}$ can be adjusted to make the prior more informative ($v^{cells}=10$).
        
        \item The latent variable $x_{rf}$ represents the contribution of each latent cell type group $r$ to the abundance of cell types $f$ and is Gamma distributed with a hierarchical prior controlling on the number of cell types $f$ that have high values of $x_{rf}$ in each group $r$:
        \begin{equation} \label{eq:c2l:10}
        x_{rf} \sim \mathtt{Gamma}(N_r^{x} / R, N_r^{x})
        \end{equation}
        where $N_r^{x}$ represents the unobserved number of cell types $f$ for each group $r$. $N_r^{x}$ is Gamma-distributed with an informative prior:
        \begin{equation} \label{eq:c2l:11}
        N_r^{x} \sim \mathtt{Gamma}(E(N^{x}), E(N^{x}) / v)
        \end{equation}
        where $E(N^{x})$ is the average number of cell types per group as provided by the user, with high values of $E(N^{x})$ indicating that each group $r$ contains many cell types $f$, and $E(N^{x})$ indicating that the spatial abundance of each cell type $f$ is independent from other cell types. Under this prior $E(\sum_{f} z_{sr}) = 1$.
    
        \end{itemize}
    
    \item \textbf{Gene-specific multiplicative scaling factor.} $m_{g}$ is modelled as Gamma-distributed (Eq. \ref{eq:c2l:12} - \ref{eq:c2l:14}) with hierarchical prior $\alpha^M$ and $\beta^m$ reflecting the prior belief about the difference in sensitivity of single cell and spatial technologies. Specifically average change in sensitivity $\mu$, $\sigma$ with which sensitivity of individual genes deviate from $\mu$, and our uncertainty in these priors $z$ expressed as mean / variance ratio. These parameters can be automatically chosen by computing average difference in gene expression mean between single cell and spatial dataset.
    \begin{equation} \label{eq:c2l:12}
    m_{g} \sim \mathtt{Gamma}(\alpha^m, \beta^m) 
    \end{equation}
    \begin{equation} \label{eq:c2l:13}
    \alpha^m \sim \mathtt{Gamma}(\mu ^ 2 / \sigma ^ 2,  \sqrt{\mu ^ 2 / \sigma ^ 2 / z})
    \end{equation}
    \begin{equation} \label{eq:c2l:14}
    \beta^m \sim \mathtt{Gamma}(\mu / \sigma ^ 2,  \sqrt{\mu / \sigma ^ 2 / z})
    \end{equation}
    
    $\mu ^ 2 / \sigma ^ 2$ converts our guess of the average change in sensitivity $\mu$ into the mean of the shape parameter of $\alpha^m$. This way we are not forcing change in sensitivity $\mu$ but allowing the model to learn it. Rate of Gamma distribution $\beta^m$ is defined according to Eq. \ref{eq:c2l:14}.
    
    \item \textbf{Additive background for genes and locations.} $l_s$ and $s_{g}$ are modelled as Gamma distributed with hierarchical priors on shape and rate of their distributions eq. \ref{eq:c2l:15} - \ref{eq:c2l:20}:
    \begin{equation} \label{eq:c2l:15}
    l_s \sim \mathtt{Gamma}(\alpha^l,  \beta^l)
    \end{equation}
    \begin{equation} \label{eq:c2l:16}
    \alpha^l \sim \mathtt{Gamma}(1, 1)
    \end{equation}
    \begin{equation} \label{eq:c2l:17}
    \beta^l \sim \mathtt{Gamma}(1, 1)
    \end{equation}
    
    \begin{equation} \label{eq:c2l:18}
    s_{g} \sim \mathtt{Gamma}(\alpha^s,  \beta^s)
    \end{equation}
    \begin{equation} \label{eq:c2l:19}
    \alpha^s \sim \mathtt{Gamma}(1, 1)
    \end{equation}
    \begin{equation} \label{eq:c2l:20}
    \beta^s \sim \mathtt{Gamma}(1, 1)
    \end{equation}
    
    \item \textbf{Overdispersion for each gene.} Containment prior is used for modelling unobserved variance using NB overdispersion $\alpha_g$ \cite{simpson_penalising_2017}. Under the following prior most genes have low overdispersion:
    \begin{equation} \label{eq:c2l:21}
    \alpha_g = 1 / o_g ^ 2
    \end{equation}
    \begin{equation} \label{eq:c2l:22}
    o_g \sim \mathtt{Exponential}(\mu^o)
    \end{equation}
    \begin{equation} \label{eq:c2l:23}
    \mu^o \sim \mathtt{Gamma}(mu,  sd)
    \end{equation}
    Constants $mu$ and $sd$ parameterising Gamma distribution in eq \ref{eq:c2l:23} are chosen to appropriately scale the exponential distribution in eq \ref{eq:c2l:22}. The inference of this parameter appears to be fairly robust to the choice of $mu$ and $sd$.\newline

\end{enumerate}


In addition to absolute cell density we also compute absolute number of mRNA molecules ($u_{sf}$) contributed by each cell type $f$ to each location $s$:
\begin{equation} \label{eq:c2l:24}
u_{sf} = w_{sf} (\sum_{g} {m_{g} \: g_{fg}})
\end{equation}
These parameters are referred to in the package as \mintinline{python}|'nUMI_factors'|. $u_{sf}$ are a more robust estimate than either absolute cell density $w_{sf}$ or relative cell proportions. \newline

\begin{figure}
    \includegraphics[scale=0.35]{images/CoLocationModelNB4V2.png}
    \caption{Illustration of the corresponding graphical model.}
    \label{fig:graphical_model}
\end{figure}

\subsection{Multi-experiment extension of cell2location model} \label{c2l_multi}

When analyzing multiple spatial data sets jointly, where $e=\{1..E\}$ denote individual experiments, such as 10X Visium square capture area and Slide-Seq V2 "puck", the over-dispersion parameter $\alpha_{e,g}$ and additive background parameter $s_{e,g}$ are fit separately for each experiment as well as genes. Specifically, $\alpha_{e,g}$ depends on $o_{e,g}$ which is modelled as independent and exponentially distributed (eq \ref{eq:c2l:21multi} and \ref{eq:c2l:22multi}:
    \begin{equation} \label{eq:c2l:21multi}
    \alpha_{e,g} = 1 / o_{e,g} ^ 2
    \end{equation}
    \begin{equation} \label{eq:c2l:22multi}
    o_{e,g} \sim \mathtt{Exponential}(\mu^o)
    \end{equation}

Individual locations belong to experiments, $s \in e$, therefore the only following parameters are shared between experiments:
\begin{enumerate}

    \item The same reference cell type signatures $g_{f,g}$ are used for all experiments.
    
    \item $m_{g}$, the change in sensitivity between technologies is shared across experiments, improving the ability of the model to distinguish low $m_{g}$ from zero cell abundance $w_{r,f}$. This is equivalent to regressing out the effect of technology but not the effect of individual experiment. 
    
    \item $x_{r,f}$, latent groups $r$ of reference cell types $f$ with similar spatial abundance, and it's hierarchical prior, $N_r^{x}$, the number of cell types per group
    
    \item Variables that serve as hierarchical priors (single value for each variable): $\alpha^m$, $\beta^m$, $\alpha^l$, $\beta^l$, $\alpha^s$, $\beta^s$, $\mu^o$
    
\end{enumerate}

\begin{equation} \label{eq:c2l:1multi}
D_{s,g} \sim \mathtt{NB}(\mu_{s,g}, \alpha_{e,g}) \\
\end{equation}

\red{\< Oli's suggestion start (eq 27):}
\begin{equation} \label{eq:c2l:1multiO}
D_{s,g}^{e} \sim \mathtt{NB}(\mu_{s,g}^{e}, \alpha_{g}^{e}) \\
\end{equation}

\subsection{Inference} \label{c2l_inference}

Variational Bayesian Inference is used to approximate the posterior, specifically black-box Automatic Differentiation Variational Inference (ADVI) implemented in pymc3 framework \cite{salvatier_probabilistic_2016}. Appropriately transformed (to positive scale) univatiate normal distributions are used to approximate each parameter. Inference is achieved by maximising log-likehood of the data and minimising KL divergence from the posterior to prior (ELBO loss function).
Several restarts are performed to evaluate the stability of the inferred posterior distribution (usually 2). Under default parameters inference is very stable so no selection of the best model is performed (the first training restart is used). Validity of selected hyper-parameters is evaluated using prior predictive check. Performance of the training is evaluated by performing the posterior predictive check for the accuracy of the model at reconstructing expression levels of each gene $g$ in each location $s$; and by evaluating consistency of inferred $w_{sf}$ parameters between the training restarts.

\subsection{Processing of the spatial expression matrix} \label{c2l_sp_proc}
Expression $D_sg$ counts of each gene $g$ in each location $s$ are obtained from running 10X Spaceranger 1.0.0 on sequencing data from the 10X Visium experiment described in Section \ref{data_analysis_mouse_brain}. Untransformed and unnormalised count matrix was filtered to include genes expressed in the single cell reference $g_fg$ and used as input to the model.

\subsection{Estimation of reference expression signatures of cell types} \label{c2l_ref_prog}

We obtain reference expression signatures $G$ from scRNA-seq reference expression data $J=\{d_{s,g}\}$ of each gene $g=\{1..G\}$ in cell $c=\{1..C\}$. Untransformed and unnormalised count matrix $J_{cg}$ is filtered to select expressed genes as described in \ref{data_analysis_mouse_brain}. Using matched scRNA-seq profiles allows to know exactly which cells express which genes in the spatial sample. Using unmatched reference instead infers the location of the most similar regulatory program (note the difference in interpretation). We use following 2 alternative methods to compute $g_{fg}$:
\begin{enumerate}
    \item Analytically computing single cell reference cluster centroids for each gene $g$ and cell cluster $f=\{1..F\}$. This method is very computationally efficient and gives very good mapping quality (measured on simulated data) when the scRNA-seq reference is a single sample or a few samples matching the source of spatial data $D_{sg}$ with no strong batch effect between them.
    \item To address the usage of challenging reference data composed of multiple batches $e=\{1..E\}$ and technologies $t=\{1..T\}$ the expression of each gene $g$ in each single cell reference cluster $f$ is inferred using a regularised Negative Binomial regression. The model accounts for the sequencing coverage $h_e$, difference in expression level between technologies $p_{tg}$ and background expression $b_{eg}$ in each sample $e$. It models unexplained variance (overdispersion $\alpha_g$) and count nature of the data using Negative Binomial distribution:
    \begin{equation} \label{eq:c2l_ref_prog:1}
    J_{cg} \sim \mathtt{NB}(\mu_{cg}, 1 / \alpha_g^2)
    \end{equation}
    \begin{equation} \label{eq:c2l_ref_prog:2}
    \mu_{cg} = (g_{fg} + b_{eg}) \: {h_e} \: p_{tg}
    \end{equation}
    All model parameters are constrained to be positive to simplify interpretation. Weak L2 regularisation of $g_{fg}$ / $b_{eg}$ / $\alpha_g$ and penalty for large deviations of $h_e$ and $p_{tg}$ from 1 is used. $g_{fg}$ is initialised at analytical average for each $f$ and $b_{eg}$ is initialised at average expression of each gene $g$ in each sample $e$ divided by a factor of 10. The informative initialisation leads to fast convergence.  \newline
    Pytorch implementation of training using mini batches makes model scalable to more than 100k cells and fast (30sec-5min on GPU). Maximum a posteriori optimisation is used to find parameters of this model with batch size of 1024 cells, ADAM optimiser learning rate 0.01. The number of epochs needed for training is determined using cross-validation performed on held-out 10 percent of cells. \newline
    To verify that the model successfully accounted to non-biological sample effects $h_e$, $p_{tg}$ and $b_{eg}$ a corrected the expression matrix $J^{corrected}_{cg}$ is obtained from $J_{cg}$, followed by standard scanpy workflow {\cite{wolf_scanpy_2018} to verify the mixing of the cells from matching cells types but different samples and technologies:
    \begin{equation} \label{eq:c2l_ref_prog:3}
    J^{corrected}_{cg} = J_{cg} / ({h_e} \: p_{tg}) - b_{eg}
    \end{equation}

\end{enumerate}

\subsection{Differential expression between reference signatures} \label{c2l_ref_prog_diff_expression}

We obtain reference expression signatures $G$ from scRNA-seq reference expression data $J=\{d_{s,g}\}$ of each gene $g=\{1..G\}$ in cell $c=\{1..C\}$. Untransformed and unnormalised count matrix $J_{cg}$ is filtered to select expressed genes as described in \ref{data_analysis_mouse_brain}.
To address the usage of challenging reference data composed of multiple batches $e=\{1..E\}$ and technologies $t=\{1..T\}$ the expression of each gene $g$ in each single cell reference cluster $f$ is inferred using a regularised Negative Binomial regression. The model accounts for the sequencing coverage $h_e$, difference in expression level between technologies $p_{tg}$ and background expression $b_{eg}$ in each sample $e$. It models unexplained variance (overdispersion $\alpha_g$) and count nature of the data using Negative Binomial distribution:
\begin{equation} \label{eq:c2l_ref_prog_diff:1}
J_{cg} \sim \mathtt{NB}(\mu_{cg}, 1 / \alpha_g^2)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:2}
\mu_{cg} = (g_{fg} + b_{eg}) \: {h_e} \: p_{tg}
\end{equation}

All model parameters are constrained to be positive to simplify interpretation. The following priors are used:
\begin{equation} \label{eq:c2l_ref_prog_diff:3}
g_{fg} \sim \mathtt{Gamma}(gs, gr)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:4}
gs \sim \mathtt{Gamma}(mu=0.3, var=mu/3)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:4}
gr \sim \mathtt{Gamma}(mu=0.8, var=mu/3)
\end{equation}

\begin{equation} \label{eq:c2l_ref_prog_diff:5}
b_{eg} \sim \sim \mathtt{Gamma}(bs, br)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:6}
bs \sim \mathtt{Gamma}(mu=0.45, var=mu/3)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:7}
br \sim \mathtt{Gamma}(mu=11, var=mu/3)
\end{equation}

\begin{equation} \label{eq:c2l_ref_prog_diff:8}
h_e \sim \mathtt{Gamma}(6, 6)
\end{equation}
\begin{equation} \label{eq:c2l_ref_prog_diff:9}
p_{tg} \sim \mathtt{Gamma}(10, 10)
\end{equation}

$g_{fg}$ is initialised at analytical average for each $f$ and $b_{eg}$ is initialised at average expression of each gene $g$ in each sample $e$ divided by a factor of 10. The informative initialisation leads to fast convergence.  \newline
Pyro implementation of training using mini batches makes model scalable to more than 100k cells and fast (5-10min on GPU). Optimisation is used to find Variational Posterior of the parameters with batch size of 1024 cells, ADAM optimiser learning rate 0.001. The number of epochs needed for training is determined using cross-validation performed on held-out 10 percent of cells. The training is stopped before $p(J_{cg} | \mu_{cg}, 1 / \alpha_g^2$) starts decreasing. \newline

To find genes shared by all sub-types of cortical and thalamic astrocytes, as well as genes shared by astrocytes and neurones from the same region we applied the following algorithm to the posterior distrubution of $g_{fg}$:


\subsection{Nuclei Segmentation} \label{c2l_segmentation}

To perform nuclei segmentation, H&E-stained image of the Visium-profiled tissue was used. The image was acquired using Hamamatsu slide scanner as described in the section on experimental processing. Full resolution images were divided into smaller parts of size about 1000 by 1000 pixels. Then a state-of-the-art convolutional neural network (CNN)-based segmentation pipeline described in \cite{caicedo_nucleus_2019} was used and available at \cite{'https://github.com/selimsef/dsb2018_topcoders'}. An ensemble of 32 pre-trained CNNs each with Unet- or FPN-like architecture was used to classify pixels to 3 classes: background, nuclei and nuclei boundaries. Predicted segmentation masks were averaged across CNNs, then a watershed algorithm was applied to refine predicted boundaries and separate individual nuclei. As a final step, a gradient boosting model over morphological features such as nuclei colour, shape and size was used to remove false-positively detected nuclei.  
Obtained individual nuclei masks were used to compute several morphological features such as area (count of pixels in a mask), shape (ratio of lengths of the major and minor axis of an ellipse fitted to the mask) and position (center of mass of the image containing individual mask).
Kd-tree approach was used to assign nuclei locations to Visium spots efficiently. Kd-tree implementation from the scikit-learn package was used.


\section{Clustering of tissue regions and diffusion maps of cell type gradients} \label{auto_clustering}

The cell density $w_{sf}$ (eq. \ref{eq:c2l:4}) inferred by cell2location model serves as a proxy to cell density. This enables describing similarity of locations (spots) by constructing a KNN graph that represents connectivity of locations in terms of their in cell composition (N neighbours = 20 for human lymph node, 38 for the mouse brain). Standard implementation of Leiden clustering (function scanpy.tl.leiden) and diffusion maps (function scanpy.tl.diffmap) in scanpy package \cite{wolf_scanpy_2018} can utilise cell composition KNN graph to characterise organisation of the tissue. We used both methods with default arguments (resolution parameter of Leiden clustering = 1). Obtained regions were cross-referenced with the mouse brain anatomy using corresponding histology images and similar clusters were merged to generate the broad region map.

\section{Combinations of co-located cell types} \label{cell_neighbourhoods}

To map out cellular interaction neighbourhoods we combined de-novo non-negative matrix factorisation model with receptor-ligand analysis using CellPhoneDB \cite{vento-tormo_single-cell_2018}. We used hierarchical Poisson matrix factorisation model to simultaneously infer combinations of co-located reference signatures and where in the tissue they locate. We assume that correlation in density of cell types across locations reflects the interaction between these cell types, specifically, any interaction that could affect mutual location of these cell types:
\begin{itemize}
    \item Signals promoting proliferation and survival of other cell type
    \item Adhesion molecules that increase the frequency of contacts 
    \item Recruitment signals such as CXCR5 / CXCL13 \cite{van_de_pavert_chemokine_2009}
    \item Repulsive signals, inhibiting the action of chemotaxis signals such as PD-1 / PD-L1 interactions \cite{shi_pd-1_2018}
    \item Interactions that induce new transcriptional states 
\end{itemize}

Visium technology serves as a native binning strategy to identify regions of action for local paracrine signals (within the $55\mu m$ spot diameter). We assume that the signal does not propagate between neighbouring spots spaced by $45\mu m$ gaps.

Therefore our Poisson factorisation model reveals which cell types interact with consequences to their location and where these interactions occur. 

To find a small number of co-located combinations representing the broad grouping of cell types - a few factors were selected. This has the advantage of identifying overlapping but distinct combinations compared to Leiden clustering.
To find very strong co-location signal, assuming that most cell types (reference signatures) are independent, a lot of factors were used (e.g. 32).

Inference is performed using standard NMF in the scikit-learn \cite{scikit-learn}.

\subsection{Description of the co-located combination model} \label{cell_neighbourhoods_model}

Density $w_{sf}$ of each cell type $f$ across locations $s$ is modelled as an additive function of the cell combinations (micro-environments) $r$. This means the density of one cell type in one location can be explained by 2 distinct combinations $r$. Cell type density is therefore a function of the following non-negative components:  
\begin{equation} \label{eq:circ:1}
w_{sf} = (\sum_{r} {i_{sr} \: k_{rf} \: m_{f}})
\end{equation}

\begin{enumerate}
    \item $k_{rf}$ represents the proportion of cells of each type (reference signatures) $f$ that correspond to each co-located combination $r$, normalised for total abundance of each cell type $m_{f}$.
    \item $m_{f}$ cell type budget accounts for the difference in abundance between cell types, thus focusing the interpretation of $k_{rf}$ on cell co-location.
    \item $i_{sr}$ is proportional to the number of cells from each neighbourhood $r$ in each location $s$, and shows the abundance of combinations $r$ in locations $s$.
\end{enumerate}

In practice $q_{rf} = k_{rf} \: m_{f}$ is obtained from scikit-learn NMF and normalised by the sum across combinations $r$ to obtain $k_{rf}$:
\begin{equation} \label{eq:circ:2}
k_{rf} = q_{rf} / (\sum_{r} q_{rf})
\end{equation}

\section{Data sources and analysis} \label{data_analysis}

\subsection{Matched Visium and single nucleus reference data of the mouse brain} \label{data_analysis_mouse_brain}

To generate matched Visium and single nucleus reference data of the mouse brain, the brains of two mice were dissected and sectioned to obtain the full coronal sections of the region containing the hippocampus. These tissue blocks were OCT embedded and fresh-frozen to enable sectioning. Spatial data was successfully generated from three 10$\mu m$ sections from mouse #28 and two from mouse #29 using 10x Visium protocol. Matching single nucleus data was generated from adjacent 50-200$\mu m$ sections (3 sections per mice) using 10x 3' v3 Chromium protocol. Both technologies yield libraries that were sequenced using short read Illumina technology by Sanger Institute sequencing facility. Sequencing data processed using 10X CellRanger (snRNA-seq) and 10X SpaceRanger (Visium spatial) into count input matrices by Sanger Institute support team (Stephen Leonard).

SnRNA-seq data was processed using standard Seurat V3 workflow without correcting batch effects between 6 invidual samples (scale-factor normalisation, log1p-transform, selecting highly variable genes, performing PCA, KNN construction, UMAP). To generate comprehensive cell annotation joint Louvain clustering was performed on all samples and clusters were subsequently annotated using marker genes from the literature and existing mouse brain single cell references \cite{zeisel_molecular_2018, tasic_shared_2018}. 

Visium spatial sequencing data and the histology image were processed using 10X SpaceRanger which selects Visium spots covered by tissue by aligning know spot locations with fudicial border spots in the histology image. This allows evaluating the correspondence between cell maps produced using our method and the known brain anatomy. This also allows identifying the number of nuclei in each spot using nuclear segmentation as described in section \ref{c2l_segmentation}. The histology image was used to manually annotate the primary Somatosensory cortex (SSp) cortex (SSp) region.

\subsection{Matched Visium and single nucleus reference data of the mouse brain} \label{validation_mouse_brain}

For validating the cell type maps produced by the cell2location model, publicly available \cite{yao_taxonomy_2020} gold stardard reference of cell location was used. In that study individual layers of the cortex are dissected and subjected to scRNA-seq using SmartSeq 4 protocol. Data from that study was obtained from Allen institute data portal \cite{https://brain-map-portal-stage.us.aldryn.io/atlases-and-data/rnaseq}. Cells originating from the primary primary Somatosensory cortex (SSp) were selected, and all cell types contributing less then 3 cells were filtered. This data was used to compute the proportion of each reference cell type in each cortical layer. 

\subsection{Unmatched Visium data of the human lymph node and single cell reference data of lymphoid organs} \label{data_analysis_LN}

Visium spatial transcriptomics data was downloaded from the 10X Genomics website. Single cell reference was combined from 3 studies by King et al \cite{king_antibody_2020, james_distinct_2020, park_cell_2020}. Integration was performed using the standard workflow of regressing out batch from PC space and performing BBKNN. The integrated reference was used to harmonise cell type and subtype labels between datasets. These labels were used to estimate reference signatures used for mapping as described in Section \ref{c2l_ref_prog}.

\printbibliography

\end{document}